### PR TITLE
feat: variants support

### PR DESCRIPTION
**Pull request by mirkosoft**
Enable variants support in Spartacus; currently apparel accelerator logic has been re-applied here with just minor UX/SEO improvements.

Works OOTB with apparel setup (style & size) and electronics (color) sampledata delivered with b2c accelerator hybris extensions. 

CMS component *ProductVariantSelectorComponent* was used to implement variants so to have this feature enabled in a custom app component should be mapped similarly to other CMS components:

```typescript
ConfigModule.withConfig(<CmsConfig>{
      cmsComponents: {
        ProductVariantSelectorComponent: {
          component: ProductVariantSelectorComponent,
          guards: [ProductVariantGuard],
        },
      },
    }),
```
For a custom variant setup, product model as well as the component's template needs to be extended - check the `product.model.ts` and `product-variant-selector.component.html` files to see how it was done for apparel/electronics. For custom variant selection templates, new components might be created and called from the main variants component mentioned, similarly to the default style, size & color selector components.

closes: #3794

### CODE REVIEW

File: projects/core/package.json [hunk 1]
```diff
@@ -16,9 +16,9 @@
   "peerDependencies": {
     "@angular/common": "^8.0.0",
     "@angular/core": "^8.0.0",
-    "@ngrx/effects": "^8.0.1",
-    "@ngrx/router-store": "^8.0.1",
-    "@ngrx/store": "^8.0.1",
+    "@ngrx/effects": "^8.3.0",
```
Comment (dunqan) [inline projects/core/package.json]: What's the reason for this?

File: projects/core/src/occ/adapters/product/default-occ-product-config.ts [hunk 2]
```diff
@@ -5,7 +5,7 @@ export const defaultOccProductConfig: OccConfig = {
     occ: {
       endpoints: {
         product:
-          'products/${productCode}?fields=DEFAULT,averageRating,images(FULL),classifications,manufacturer,numberOfReviews,categories(FULL)',
+          'products/${productCode}?fields=FULL,averageRating,images(FULL),classifications,manufacturer,numberOfReviews,categories(FULL)',
```
Comment (dunqan) [inline projects/core/src/occ/adapters/product/default-occ-product-config.ts]: Can we have something more specific than `FULL` here? We are moving away from using `FULL`, 'DEFAULT' and 'BASIC' anyway.

File: projects/storefrontlib/src/cms-components/cart/cart-shared/cart-item/cart-item.component.html [hunk 3]
```diff
@@ -29,11 +29,15 @@
...
+          *ngFor="
+            let variant of item.product.baseOptions[0]?.selected
+              .variantOptionQualifiers
+          "
           class="cx-property"
         >
-          <div class="cx-label">{{ variant.name }}</div>
-          <div class="cx-value">{{ variant.value }}</div>
+          <div class="cx-label" *ngIf="variant.name && variant.value">
+            {{ variant.name }}: {{ variant.value }}
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/cart/cart-shared/cart-item/cart-item.component.html]: DOM structure change is considered a breaking change.

File: projects/storefrontlib/src/cms-components/product/product-variant-selector/size-selector/size-selector.component.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    private routingService: RoutingService
+  ) {}
+
+  @Input()
+  product: Product;
+
+  @Input()
+  variants: BaseOption;
+
+  baseUrl = this.config.backend.occ.baseUrl;
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/product-variant-selector/size-selector/size-selector.component.ts]: Do we need it as a public property?
Also, for config, it's safer to use getters.

File: projects/storefrontstyles/scss/layout/page-templates/_product-detail.scss [hunk 5]
```diff
@@ -8,29 +8,29 @@
 
       @include media-breakpoint-up(lg) {
         display: grid;
-        grid-template-columns: 1fr 1fr;
-        grid-template-rows: repeat(2, auto) 1fr;
         grid-column-gap: 20px;
         grid-row-gap: 0px;
+        grid-template-columns: 1fr 1fr;
```
Comment (dunqan) [inline projects/storefrontstyles/scss/layout/page-templates/_product-detail.scss]: Change css of existing classes is considered a breaking change.

File: projects/core/src/occ/adapters/product/default-occ-product-config.ts [hunk 6]
```diff
@@ -5,7 +5,7 @@ export const defaultOccProductConfig: OccConfig = {
     occ: {
       endpoints: {
         product:
-          'products/${productCode}?fields=DEFAULT,averageRating,images(FULL),classifications,manufacturer,numberOfReviews,categories(FULL)',
+          'products/${productCode}?fields=DEFAULT,averageRating,images(FULL),classifications,manufacturer,numberOfReviews,categories(FULL),baseOptions,baseProduct,variantOptions,variantType',
```
Comment (dunqan) [inline projects/core/src/occ/adapters/product/default-occ-product-config.ts]: `variantType` is missing on product_scopes.details, and I guess it should be included

File: projects/storefrontlib/src/cms-components/product/product-list/product-grid-item/product-grid-item.component.html [hunk 7]
```diff
@@ -31,6 +31,10 @@
   </div>
 </div>
 
+<div class="cx-style-icons" *ngIf="product.variantOptions">
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/product-list/product-grid-item/product-grid-item.component.html]: Should we consider it as a breaking or not?

File: projects/storefrontlib/src/cms-components/product/product-list/product-list-item/product-list-item.component.html [hunk 8]
```diff
@@ -43,5 +43,11 @@
         ></cx-add-to-cart>
       </div>
     </div>
+    <div class="row" *ngIf="product.variantOptions">
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/product-list/product-list-item/product-list-item.component.html]: Should we consider it as a breaking or not?

File: projects/storefrontlib/src/cms-components/product/product-variant-selector/style-icons/style-icons.component.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  OccConfig,
+  VariantOption,
+  VariantOptionQualifier,
+  VariantQualifier,
+} from '@spartacus/core';
+
+@Component({
+  selector: 'cx-style-icons',
+  templateUrl: './style-icons.component.html',
+  styleUrls: ['./style-icons.component.scss'],
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/product-variant-selector/style-icons/style-icons.component.ts]: It is possible to make it with ` changeDetection: ChangeDetectionStrategy.OnPush,`?

File: projects/storefrontlib/src/cms-components/product/product-variant-selector/product-variant-selector.component.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
+import { Product, BaseOption, VariantType } from '@spartacus/core';
+import { Observable } from 'rxjs';
+import { CurrentProductService } from '../current-product.service';
+import { tap, filter, distinctUntilChanged } from 'rxjs/operators';
+
+@Component({
+  selector: 'cx-product-variant-selector',
+  templateUrl: './product-variant-selector.component.html',
+  changeDetection: ChangeDetectionStrategy.Default,
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/product/product-variant-selector/product-variant-selector.component.ts]: We don't need `ChangeDetectionStrategy.Default` probably, btw, is it possible to make it OnPush?

File: projects/storefrontstyles/scss/layout/page-templates/_product-detail.scss [hunk 11]
```diff
@@ -8,10 +8,9 @@
...
```
Comment (dunqan) [inline projects/storefrontstyles/scss/layout/page-templates/_product-detail.scss]: We still have missing `grid-template-rows: repeat(2, auto) 1fr;`, yup?

File: projects/storefrontstyles/scss/components/product/details/_product-summary.scss [hunk 12]
```diff
@@ -17,6 +17,9 @@
     margin: 0 0 11px 0;
   }
 
+  p.summary p {
+    margin-bottom: 0;
```
Comment (dunqan) [inline projects/storefrontstyles/scss/components/product/details/_product-summary.scss]: hmm... not sure about that, but it seems we are essentially targetting existing element (already present) with different styles, what do you think?

Comment (mirkosoft) [reply projects/storefrontstyles/scss/components/product/details/_product-summary.scss]: will remove that as it's better to make such adjustments using a theme

**Pull request merged**